### ISSUE TITLE
Introduce Steam Timeline support

### ISSUE THREAD

**Issue by zackman0010** [bluemagic123/tModLoader#4324]
### Do you intend to personally contribute/program this feature?

Yes

### I would like to see this change made to improve my experience with

Gameplay as a Player

### Description

I would like to integrate tModLoader with the new [Steam Game Recording](https://store.steampowered.com/gamerecording) features. Using the new Steam Timeline API, games have the option to customize the timeline of the recording by adding event markers and highlights. Terraria could use these to mark player deaths and boss kills, as well as opening the capability to modders to add their own events to the timeline.

Example screenshot from my local WIP branch:
![image](https://github.com/user-attachments/assets/0166c1ea-7d54-4aac-adf9-04d424212495)

### What does this proposal attempt to solve or improve?

This would allow tModLoader to be one of the first games to add Steam Timeline support (As of the time of my writing this, there are currently 28 games tagged with Steam Timeline on the Store).

**Comment by direwolf420** [comment]
If this ends up getting added, it's important to still have the game work normally on gog or when using a steam version where this feature is not supported

**Comment by zackman0010** [comment]
I was already testing GOG, but I hadn't considered testing older Steam versions. I'm fairly certain the API handles that internally, but I'll verify by removing my client from the Beta and retesting.

**Comment by direwolf420** [comment]
If you need help with this I run an outdated steam build (march 2023)

**Comment by zackman0010** [comment]
Sure! I'm waiting for Steamworks.NET to officially [release the next version](https://github.com/rlabrecque/Steamworks.NET/issues/620) with Timeline support before I make a WIP PR here. Right now I'm using a version I built locally to develop with.

**Comment by JavidPack** [comment]
Looks like custom icons are controlled by the game dev, so those wont be customizable by modders, but everything else should be easy enough. I'm not even sure if we need anything to support this other than updating the steamworks dll, but we could certainly add some helper methods if we want to make it simpler for modders.

The video guide in the docs seems to suggest we can check if the steam client supports this feature without running into issues, so we'd only need to wait for steamworks.net not steam client itself, I imagine.

I think it would be good to plan for how multiple mods might desire to mark the same events (boss fights, player death), leading to the timeline being too cluttered. I think our ExampleMod examples and how we present the feature to modders should encourage modders to exercise restraint, maybe suggesting that any usage should be gated by a config option at least. It remains to be seen how popular the feature is, but this reminds me of how many mods started supporting Discord Rich Presence when that was released.

**Comment by zackman0010** [comment]
My idea was to support deaths and vanilla boss kills from tModLoader, both so mods would only have to worry about adding their own events and so the vanilla bosses could have their heads uploaded for custom event icons. I'm hoping a later update to the beta will allow Workshop mods to upload their own icons as well.

**Comment by JavidPack** [comment]
> My idea was to support deaths and vanilla boss kills from tModLoader, both so mods would only have to worry about adding their own events and so the vanilla bosses could have their heads uploaded for custom event icons. I'm hoping a later update to the beta will allow Workshop mods to upload their own icons as well.

That sounds like it might be overstepping a bit for tModLoader itself to do, but we can decide on that later once we get feedback from others. I think tModLoader itself should just do the main menu and in-game timeline support.

**Comment by Solxanich** [comment]
```steam client supports this feature without running into issues```

Note that the Init method was changed for this, so be careful when upgrading that we explicitly verify SyeamAPI.init and GameServer.init

**Comment by zackman0010** [comment]
So I've run into a small roadblock while working on this. It appears that the Timeline recording locks onto the first process that runs `SteamAPI.Init()`. For most games, this isn't an issue, but tModLoader starts a separate process in `TerrariaSteamClient`. I assume this is so playtime is tracked for both vanilla Terraria and tModLoader. The problem here is that the Timeline tries to lock onto this process, meaning it's not recording any actual gameplay.
![image](https://github.com/user-attachments/assets/fee797fe-978e-4d96-ab16-1809a806d81b)
Fortunately, the "Recording other game" is actually a button you can press to lock Timeline onto the correct process, but as it is the player will have to remember to go into the Steam overlay and press that button every time they launch the game.

From what I can see, there's no real good solution here. For Timeline to lock onto the correct process, tModLoader has to be registered first. But if tModLoader is registered first, then registering Terraria will make Steam's "Now Playing" show Terraria instead of tModLoader. From what I've been able to tell, the only way to have both Timeline automatically locked onto the correct process and the Now Playing show tModLoader would be to allow TerrariaSteamClient to close after it runs the install validations, but then it wouldn't be tracking any playtime on vanilla Terraria.

**Comment by JavidPack** [comment]
Yes, that is a bit of an issue. The playtime tracking is a very complex situation that would be prone to breaking if we tried to reorder things. We'll have to keep looking into this, maybe there is a solution that works, but tModLoader is a bit unique in how it interacts with the Steam API, so this might come down to a compromise. I think we intend to keep the playtime tracking functional as it is a higher priority, but we'll need to revisit it once again internally and see if it's working as intended.

**Comment by zackman0010** [comment]
@JavidPack Let me know if there's anything I can do to help! Unfortunately I don't have the collaborators role in Discord yet, so I'm not able to participate there, but I'm watching this thread.

The issue with the timeline locking onto the wrong process can be recreated even without my WIP branch. The branch I'm working on is for the update of Steamworks.NET and the enhanced integration (coloring the timeline based on being in the menu/in-game, event icons), but the base Timeline functionality can still be used without any changes on the game's side.

Timeline is also still a beta feature, and tModLoader actually isn't the only game that has this problem. Looking on the forums, there's reports of games that have EasyAntiCheat enabled having Timeline lock onto the EAC window instead of the game. So Valve may provide a way to fix this in the future.

**Comment by JavidPack** [comment]
Anyone who wants the collaborators role can get it, just ping me (jopojelly) on discord and we can start a thread there. (I didn't see a user with your github username)

It might be worth just waiting and seeing if steam has a solution, or we can leverage our contacts to see that something gets done about it. Lets talk more there, I think other contributors will have more to say about the expected default implementation and they are more likely to talk there than here.
